autoprotocol/builders.py

```diff
@@ -1738,6 +1747,81 @@ def blowout(self, volume, initial_z, flowrate=None):
             "flowrate": flowrate
         }
 
+    def mode(self, transports=None, mode=None):
+        """Helper for selecting dispense mode based on liquid_class name
+
+        Parameters
+        ----------
+        transports : dict, optional
+            Dictionary of the transport parameters
+        mode : str, optional
+            Mode of dispense type
+
+        Returns
+        -------
+        str
+            Mode of dispense type
+
+        Raises
+        ------
+        ValueError
+            mode does not contain valid mode value
+        TypeError
+            liquid is not of str
+        ValueError
+            liquid_class is not a valid value
+        ValueError
+            multiple liquid_class exists in one LiquidHandle
+        """
+        liquid_classes = set([transport["mode_params"]["liquid_class"]
+                              for transport in transports])
+        # remove automatically added 'air' class from blowout, etc.
+        non_air_classes = [liq for liq in liquid_classes if liq != "air"]
+        if mode:
+            if mode not in self.dispense_modes:
+                raise ValueError(
+                    "mode: {} must be one of the valid modes: {}"
+                    "".format(mode, self.dispense_modes)
+                )
+        modes = []
+        # get mode for each liquid_class if not specified already from the
+        # user input.
+        for liquid in non_air_classes:
+            if liquid is not None:
+                if not isinstance(liquid, str):
+                    raise TypeError(
+                        "liquid: {} is not of type str or None".format(liquid)
+                    )
+                if liquid not in self.liquid_classes:
+                    raise ValueError(
+                        "liquid_class: {} must be one of the valid classes: {} "
+                        "".format(liquid, self.liquid_classes)
+                    )
+                if mode is None:
+                    if liquid == "viscous":
+                        modes.append("positive_displacement")
+                    else:
+                        modes.append("air_displacement")
+            else:
+                if mode is None:
+                    modes.append("air_displacement")
+        # return error if there are incompatible liquid_class in one set of
+        # transports.
+        if len(set(modes)) > 1:
+            raise ValueError(
+                "There are multiple liquid classes which could potentially"
+                "have different modes: {}."
+                "Please specify the mode to be used from: {}."
+                "".format(modes, self.dispense_modes)
+            )
+        # if all liquid classes happen to be "air", mode is "air_displacement"
+        if not modes:
+            mode = "air_displacement"
+        else:
+            mode = list(modes)[0]
```

Comment:
I'd add a few more conditionals to check that the 0th item in this list really is the mode you want. I think including this. Maybe moving this closer to the conditional where you check the length of set of modes is 1 would be a good place for it, then `if len(set(modes)) > 1`, raise that error?

Reply:
I see. I'll update.